### DISCUSSION TITLE
Text display mechanisms in AIR

### DISCUSSION THREAD

**Discussion by ajwfrost** [airsdk/Adobe-Runtime-Support#3109] [General]
Hi

We've been looking at the text output options in AIR, and how the various different options work across platforms - with a particular focus on how to achieve new(ish) things like support for coloured emoji characters.

There are a lot of different ways of doing text, so trying to put some information down here, and see what people's experiences are and what new requirements there might be....

----

One quick note before we dive into the text display: some information about the `String` class itself. This is implemented in a similar way to the JavaScript string class, which means it should be UTF-16. In fact there seems to be a lot of assumption in the code that it's UCS-2 i.e. two-byte characters; support for code points above 0xFFFF was added later but only partly. So, if you (up to 51.0.0.3) call `String.fromCharCode(value)`, then `value` will be truncated to 16 bits. There is handling with the textblock related code (Flash Text Engine) to deal with surrogate pairs i.e. the UTF-16 encoding of a Unicode code point higher than 0xFFFF, and we've now corrected `String.fromCharCode` so that if you call it with a higher value - like emoji 0x1F600 - it will return a string that contains two character codes, which together represent that emoji.

So, be aware that if you're dealing with some high code point values, the string length will give you the number of UTF-16 entries, but conceptually that may not equal the number of characters you'll see when displayed. In fact, it's possible to have e.g. 5 UTF-16 entries which should be displayed only as a single character, if you look at modifiers..

Also fyi the `String` class does have optimisations for when it deals only with ascii characters, and it should properly handle UTF-8 input as well e.g. when reading from byte arrays (or IDataInput) etc.

----

So: fundamentally, there are three different ways to display text:
* 'traditional' text field i.e. [flash.text.TextField](https://airsdk.dev/reference/actionscript/3.0/flash/text/TextField.html)
* 'advanced' text block/line i.e. [flash.text.engine.TextBlock](https://airsdk.dev/reference/actionscript/3.0/flash/text/engine/TextBlock.html) - referred to as FTE.
* stage text for Android/iOS i.e. [flash.text.StageText](https://airsdk.dev/reference/actionscript/3.0/flash/text/StageText.html)

`StageText` we can cover easily: on desktop platforms this just uses a normal `TextField` object; on Android/iOS it uses the native text display/input classes, rendered as an overlay layer above the AIR stage. It only supports device fonts i.e. you give it a font name and it tries to get that font from the OS, as if you were creating a native mobile application... But, because of how this works, it means that you can pass in an emoji character code and the OS will display this using its internal fallback mechanism, resulting in a properly colourful emoji.

For the other two mechanisms, they are quite different, but they both have two different ways to pick up the required font information:
1) Device fonts
2) Embedded fonts

### Device fonts

Device fonts are identified by the name, and then if those are available on the device, they are used for rendering the text. We use operating-specific mechanisms for rendering the strings, so how the fonts (and emoji characters) appear is dependent on that. So for example, on Windows we use (currently) GDI "TextOut" type methods, and these can display emojis but only in black and white. We're looking to switch to Direct2D text output, which should then support the coloured emojis.

If you specify a font such as "Arial", and then your string has an emoji character, then the font will need to use a "fallback" mechanism to find the glyph to use in displaying this character. That seems to work with TextField (?) .. although on iOS it seems that the mechanism we use for text output just doesn't support any emojis (which are bitmap-based characters, rather than outline-based). With FTE, there are a set of fallback fonts that are specified within the AIR build/configuration, based on the character categories. Currently (up to 51.0.0.3) there was no category for the emoticon range so we are adding this, using the font names that appear to be the appropriate default values for each operating system. This means that emojis should start to appear if using device text in FTE/TextBlock .. but with the earlier caveats about the OS-specific rendering. It would equally have been possible to support these by creating a `FontDescription` using the OS-specific font names such as "Segoe UI Emoji" etc.

One other difference between the traditional TextField and the FTE methods is where FTE handles advanced layout, glyph combinations, right-to-left, etc .. whereas the TextField text display just throws the string at the operating system.

### Embedded fonts

Embedded fonts are where the font definitions are built into a SWF file. There are two main variants of how a font definition is embedded:
1) Shapes (for normal TextField) - "embedded" type, using DefineFont/DefineFont2/DefineFont3 SWF tags
2) CFF outlines (for FTE) - "embeddedCFF" type, using DefineFont4 and the OpenType font format with CFF

To embed the font, you can use Animate or you can use an embed directive with the ActionScript compiler, e.g.:
```
		[Embed(source="c:\\windows\\fonts\\seguiemj.ttf", embedAsCFF="true", fontFamily="Emoji")]
		private static const Emoji : Class;
		[Embed(source="c:\\windows\\fonts\\comic.ttf", embedAsCFF="false", fontFamily="Comic")]
		private static const Comic : Class;
```

This should result in the DefineFontX tags being added to your SWF file. Note that there can sometimes be issues with font embedding, where the TTF file could contain things not supported by the conversion tool, so if you see a null reference error or problem with "addFrame" in your compiler output, check whether it helps to switch out the TTF file with a basic one..

For embedded fonts, there are currently no fallback mechanisms. So if you want to draw an emoji character, you need to select the right font for that. This does mean that it should **always** be possible for you to display text that contains emojis, on all platforms, **but** you need to do that manually i.e. detect if the text has any emojis, and if so, switch to an appropriate font that contains the emoji glyphs.

In terms of colour .. these all store the font definitions as outlines, they do not have any support for bitmap font characters which are used by emojis. You can see something similar in Microsoft Word if you try to do an "Insert symbol" and select the Segoe UI Emoji font: the emojis look black and white in the symbol picker because that uses their outlines; but once you insert the character, it renders with the bitmap and shows up in colour.

So to support coloured emojis properly, using embedded fonts, we need two things:
1) The ability to have bitmap glyph data within the font information
2) A fallback mechanism so that you don't have to manually check and switch fonts within a string

Currently we are looking just at the Flash Text Engine for this, because things like fallback mechanisms and glyph combinations (e.g. where you combine a 'flag' emoji with a 'rainbow' emoji to result in a single 'rainbow flag' character) are already dealt with in the underlying code. So then, we "just" need to add support for the bitmap and colour bitmap OpenType font tables... and of course, we also need a way for people to be able to specify the fallback font to be used in case a glyph isn't found in the primary one.

One restriction though: we currently don't have a way to update how the ActionScript compiler processes TTF files and embeds the data into the SWF. We are seeing whether it's possible to get hold of this, but as a workaround we will have a way to dynamically load a `FontDescription` from some bytearray data; this could either then be loaded in at runtime, or embedded in the SWF as a binary data field.

Will keep people updated here with details of progress and implementation mechanisms/examples. And we'd be interested in feedback/comments of course!

thanks

**Comment by waveofthought-code** [comment]
Nothing to add at this point, except an exuberant THANK YOU for the effort you guys have clearly put in to this advancement already.  I've gone to great lengths to get colored emoji working in my project and it's a hack at best while still only being a half measure.  I'd abandon it all in a heartbeat if this goes anywhere.

edit:  I am still curious, however, as to how MacOS and the old Pepper FlashPlayer in Chrome (PC) rendered colored emoji without any extra effort from us.   Like, what sorcery were they tapping in to that made it just work?

**Reply by ajwfrost** [reply]
They were using different APIs... the macOS one, I need to check actually, given there's quite an overlap between macOS and iOS! For Pepper, the Flash Player actually requested that Chrome do the text rendering, so all of their font handling code was then used.

We're currently finding that there are of course many different ways to embed coloured emojis into a font! First focus is on COLR tables, version 0, which is what the Windows Segoe UI Emoji font uses... will also check the Google Noto Emoji which uses bitmap glyph information rather than outlines..

thanks

**Comment by MalacTheLittle** [comment]
Since text input/output is a core and most common feature in most apps/games, any advancement effort is most welcome. Personally, I don't have a need for emojis in my current projects, but I'm happy to see that this feature and text engine in general get its turn.

What are the plans when could dynamically load font from bytearray you mentioned be realized and will there be support for subsetted and obfuscated fonts (#1448)?
Also, will you be able to add support for sub/sup html tags (#2012)?

**Reply by ajwfrost** [reply]
So, dynamically loading from a byte array is possible for the FTE mechanism (TextLine etc) - at least for OpenType fonts with CFF outlines. But the conversions that the compiler does when you embed a TrueType font aren't something we can (currently?) do in the runtime. It does look like the font rasteriser might actually support TrueType outlines though .. so that might actually work!

For TextField with embedded fonts, it uses Shape definitions to render the text, and that's again not something we can do within the runtime (never say never, but we don't have the ability currently ..) - so loading in a font file dynamically for a TextField isn't going to be possible I'm afraid.

For those other features - we'll look into them, to see if we can handle things e.g. the superscript/subscript in a similar manner to HTML (but also, just checking that you know there are also some Unicode code points for common super/sub script characters?)

**Reply by MalacTheLittle** [reply]
If FTE would have an option to dynamically load fonts it would be big step forward! Not sure what are the differences between TTF and OTF rendering/conversion mechanisms, it would definitely be great if both font types would be covered, even WOFF/WOFF2, but again OTF is a good starting point.

I am aware of Unicode super/sub characters, but most of the fonts don't have them defined and it would be more convenient to use `baselineShift` on available characters. 

1. Would it be possibile to add `baselineShift` property to `TextFormat` like it has `ElementFormat`?
2. What are your thoughts about dynamically loading subsetted and obfuscated fonts?
3. Can you give any prediction as to when those updates might be implemented?

**Comment by Devin-Licastro** [comment]
Really happy to hear you guys are looking into this. I'm using Starling/Feathers and have had pretty big problems both with colored emojis and sometimes when displaying characters in (some incorrect Chinese characters, limited RTL language support, for some reason I couldn't bold characters in some language on iOS with a device font or something like that)

**Reply by ajwfrost** [reply]
I think on iOS we have some issues with device text outputs still .. one of the team is looking at using different APIs to see what options we may have.

**Comment by itlancer** [comment]
@ajwfrost
Thanks for detailed explanation!
I just would like to highlight feature request about fonts files (via `ByteArray`) usage in run-time https://github.com/airsdk/Adobe-Runtime-Support/issues/149
And feature https://github.com/airsdk/Adobe-Runtime-Support/issues/2611
For me important to use texts in `DisplayList` and `Stage3D`. So may be new mechanisms could be implemented via some "new" text classes (not `TextField` and `TextBlock` if there some complexities). `StageText` great for some "native" interactions but unfortunately cannot be used in a lot of cases for rendering with `DisplayList`/`Stage3D` with user interactions.

**Reply by ajwfrost** [reply]
Per a response I just put in above -> we should be able to dynamically load font files for TextBlock/TextLine output but not for TextField...

We'll look at strikethrough though, similar to subscript and superscript I would hope there are some rules we can follow for this..

**Comment by Adrian-S** [comment]
We got names like "𝓐𝓵𝓮𝔁𝔂𝓪💗" and somehow it passed trough the app successfully. We had some combinations that failed when rendered in a Feathers UI `List()`. Support for better string encoding and rendering is always welcomed. 

String concatenation is slow. Recently I've concluded it's faster to write Strings to a `ByteArray` and add strings in the `ByteArray`, and than pull the entire string out. `Array.join(',')` is also slow. Most string operation abuse the garbage collector, and I consider that should be async.

**Reply by ajwfrost** [reply]
Interesting .. I know there's a lot that can go on when concatenating strings, because the avmplus code always seems to try to minimise memory usage of strings hence having different objects referencing the same buffers in various ways. I would have hoped the `Array.join` function could be pretty quick though, we can look at that.. and potentially see if we can pause the GC edge operations whilst doing string concatenation..

**Reply by Adrian-S** [reply]
Look at this beauty:
```
public function timetrace(...args):void {
    var text:String = "";
    var time:Date = new Date();
    for (var i:int = 0; i < args.length; i++) {
        if (args[i] == null) {
            args[i] = "NULL";
        }
        text += args[i] + " ";
    }
    
    var timestamp:String =
            String("0" + time.hours.toString()).substr(-2) + ":" +
            String("0" + time.minutes.toString()).substr(-2) + ":" +
            String("0" + time.seconds.toString()).substr(-2) + ":" +
            String("00" + time.milliseconds.toString()).substr(-3);
    
    trace("[", timestamp, ']', text);
}

public function optimizedTimetrace(... args):void {
    var ba:ByteArray = new ByteArray();
    var time:Date = new Date();
    
    // Construct the initial timestamp template
    ba.writeUTFBytes("[ 00:00:00.000 ]");
    
    // Insert the time directly into the ByteArray, using the positions
    ba.position = 2; // Adjust for the initial part of the template
    ba.writeByte(48 + int(time.hours / 10));
    ba.writeByte(48 + (time.hours % 10));
    ba.position += 1; // Skip ':'
    ba.writeByte(48 + int(time.minutes / 10));
    ba.writeByte(48 + (time.minutes % 10));
    ba.position += 1; // Skip ':'
    ba.writeByte(48 + int(time.seconds / 10));
    ba.writeByte(48 + (time.seconds % 10));
    ba.position += 1; // Skip '.'
    ba.writeByte(48 + int(time.milliseconds / 100));
    ba.writeByte(48 + int((time.milliseconds % 100) / 10));
    ba.writeByte(48 + (time.milliseconds % 10));
    ba.position += 2; // Skip ' ]'
    
    // Now add the arguments with spaces
    for each (var arg:* in args) {
        ba.writeUTFBytes(" "); // Space before each arg
        if(arg != null) {
            ba.writeUTFBytes(arg.toString());
        } else {
            ba.writeUTFBytes("NULL");
        }
    }
    
    // Calculate the total length written so far
    var totalLength:int = ba.length;
    
    // Reset position and read the entire string
    ba.position = 0;
    var text:String = ba.readUTFBytes(totalLength);
    
    trace(text);
}
```
`optimizedTimetrace()` is 2x faster than `timetrace()`, good thing this is only in debugging, but I do have the date to `Ymd` in a few places that I need to replace / remove.

Adobe Scout is still helping a lot to determine bad code. Second to GC is rendering a lot of text on screen. I don't exactly know if it's a memory allocation issue or effective font rendering. I don't know what starling does under the hood, I work on an appointments app and on 2k screen you can fit a lot of appointments each with a minimum of 3 labels in 3 font styles.

**Reply by 2jfw** [reply]
Small side-note that function `timetrace` is actually not well-optimized as it accesses the `args` array too often.
Normally, you would want to put the length to a variable and have one variable to store the accessed item of the array like so:

```
        var item : String;
	var len : uint = args.length;

	for (var i : uint = 0; i < len; i++)
	{
		item = args[i];

		if (!item)
		{
			item = "NULL"; // seems to be some kind of indication that something may be missing.. if-query should be removed in release code; maybe add conditional compilation block here
		}

		text += item + " ";
	}
```

Also the timestamp part with the multiple string operations looks a bit overheaded.
`DateTimeFormatter` (https://airsdk.dev/reference/actionscript/3.0/flash/globalization/DateTimeFormatter.html) with `DateTimeStyle`s may help you here.
